### PR TITLE
Implement Reference type

### PR DESCRIPTION
**Pull request by Sh0g0-1758**
Fixes: #645

### CODE REVIEW

**Review by cicr99 — CHANGES_REQUESTED** [review]

File: pkg/hintrunner/hinter/operand.go [hunk 1]
```diff
@@ -11,14 +11,13 @@ import (
...
-//
-// All CellRef definitions
-
-type CellRefer interface {
+type Reference interface {
 	fmt.Stringer
 
-	ApplyApTracking(hint, ref zero.ApTracking) Reference
 	Get(vm *VM.VirtualMachine) (mem.MemoryAddress, error)
+	GetAddress(vm *VM.VirtualMachine) (mem.MemoryAddress, error)
```
[the remark this answers is not in the document]
Comment (Sh0g0-1758) [reply pkg/hintrunner/hinter/operand.go]: done.

File: pkg/hintrunner/hinter/operand.go [hunk 2]
```diff
@@ -35,6 +34,22 @@ func (ap ApCellRef) Get(vm *VM.VirtualMachine) (mem.MemoryAddress, error) {
 	return mem.MemoryAddress{SegmentIndex: VM.ExecutionSegment, Offset: res}, nil
 }
 
+func (ap ApCellRef) GetAddress(vm *VM.VirtualMachine) (mem.MemoryAddress, error) {
```
[the remark this answers is not in the document]
Comment (Sh0g0-1758) [reply pkg/hintrunner/hinter/operand.go]: done.

File: pkg/hintrunner/zero/hint_reference_resolver.go [hunk 3]
```diff
@@ -35,28 +35,30 @@ func (m *hintReferenceResolver) GetReference(name string) (hinter.Reference, err
...
 	}
-	op, ok := ref.(hinter.CellRefer)
-	if !ok {
+	switch ref.(type) {
+	case hinter.Deref, hinter.DoubleDeref, hinter.BinaryOp, hinter.Immediate:
 		return nil, fmt.Errorf("expected %s to be CellRefer (got %T)", name, ref)
+	default:
+		return ref, nil
 	}
-	return op, nil
```
[the remark this answers is not in the document]
Comment (Sh0g0-1758) [reply pkg/hintrunner/zero/hint_reference_resolver.go]: done.

File: pkg/hintrunner/zero/zerohint.go [hunk 4]
```diff
@@ -391,9 +391,13 @@ func createTestAssignHinter(resolver hintReferenceResolver) (hinter.Hinter, erro
...
 
-	a, ok := arg.(hinter.ResOperander)
-	if !ok {
+	var a hinter.Reference
+
+	switch v := arg.(type) {
+	case hinter.ApCellRef, hinter.FpCellRef:
 		return nil, fmt.Errorf("expected a ResOperander reference")
+	default:
+		a = v
```
[the remark this answers is not in the document]
Comment (Sh0g0-1758) [reply pkg/hintrunner/zero/zerohint.go]: done.

File: pkg/hintrunner/zero/hintparser.go [hunk 5]
```diff
@@ -176,15 +176,33 @@ func (expression ArithExp) Evaluate() (hinter.Reference, error) {
...
+		}
+
+		var leftApResult hinter.ApCellRef
+		var leftFpResult hinter.FpCellRef
+
+		if isAp {
+			leftApResult = leftExp.(hinter.ApCellRef)
+		} else {
+			leftFpResult = leftExp.(hinter.FpCellRef)
+		}
```
[the remark this answers is not in the document]
Comment (Sh0g0-1758) [reply pkg/hintrunner/zero/hintparser.go]: done.

File: pkg/hintrunner/zero/hintparser.go [hunk 6]
```diff
@@ -176,15 +176,33 @@ func (expression ArithExp) Evaluate() (hinter.Reference, error) {
...
 	}
 
-	if leftResult, ok := leftExp.(hinter.CellRefer); ok {
-		// Binary Operation does not support CellRef in the left hand side
-		// so the expression has to follow the pattern:
-		// reg + off + off + ... + off
+	switch leftExp.(type) {
+	case hinter.ApCellRef, hinter.FpCellRef:
+		// Binary Operation does not support CellRef in the left-hand side
+		// so the expression has to follow the pattern: reg + off + off + ... + off
```
[the remark this answers is not in the document]
Comment (Sh0g0-1758) [reply pkg/hintrunner/zero/hintparser.go]: mmh can you please elaborate a little more on this. I am not sure I understand what you mean here fully.

[the remark this answers is not in the document]
Comment (cicr99) [reply pkg/hintrunner/zero/hintparser.go]: Previous to this change Binary Operation took a ResOperander as left member and a ResOperander as right one, which meant CellRef were not allowed in this operation, so an expression like the one in the comment `reg + off + off + ... + off` necessarily had to be compressed. Now, Binary Operation takes references as the two members, which means if we wanted to, we wouldn't have to compress it as we would be able to solve it either way.

I'm saying then, let's leave the compression, but there is no need to return an error if it doesn't match that expression. If it is successful it should return the compressed cellRef. After this if, always do the logic inside the else statement which will return a Binary Operation. Does it make more sense now?

[the remark this answers is not in the document]
Comment (Sh0g0-1758) [reply pkg/hintrunner/zero/hintparser.go]: Oh okay. Yeah it makes sense to me now. Made the changes.

**Review by cicr99 — APPROVED** [review]

**Pull request merged**